bnb_yaz/bnb-downloader.py

```diff
@@ -2,262 +2,425 @@
+    else:
+        # check whether the key "list_of_existing_id" exists in new_bnb_info for the file_name
+        if file_name not in new_bnb_info:
+            InitializeNewBNBInfoStructure(file_name)
+
+        new_bnb_info[file_name]["list_of_existing_id"].append(bnb_id)
+
+
+# Update date_delivered in records table
+def UpdateDateDeliveredInRecords(bnb_id):
+    cursor = db_connection.cursor()
+    # update date_delivered
+    cursor.execute("UPDATE records SET date_delivered = ? WHERE bnb_id=?", (datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"), bnb_id))
+
+
+# Get BNB IDs for downloading
+def GetNotDeliveredIDs():
+    cursor = db_connection.cursor()
+    # select bnb_id and source_file_name where date_delivered is NULL
+    cursor.execute("SELECT bnb_id, source_file_name FROM records WHERE date_delivered IS NULL")
+    rows = cursor.fetchall()
+    bnb_ids = []
+    
+    for row in rows:
+        bnb_ids.append((row[0], row[1]))
+    
+    return bnb_ids
+
+
+# Download records for a list of BNB IDs
+def DownloadRecordRange(bnb_ids):
+    global marc_output_files, new_bnb_info, number_of_downloaded_ids, the_max_records_per_batch_for_delivery
+    bnb_yaz_client.sendline("format marc21")
+    bnb_yaz_client.expect("\r\n")
+
+    counter = 0
+    source_file_name = ""
+    for bnb_id, source_file in bnb_ids:
+        # create new entry in new_bnb_info if not exist
+        if source_file not in new_bnb_info:
+            InitializeNewBNBInfoStructure(source_file)
+        
+        # group the output marc file by source file name
+        if( source_file != source_file_name):
+            source_file_name = source_file
+            marc_output_file = "bnb-" + datetime.datetime.now().strftime("%y%m%d") + "-" + source_file_name + ".mrc"
+            bnb_yaz_client.sendline("set_marcdump " + working_directory + "marc/" + marc_output_file)
+            bnb_yaz_client.expect("\r\n")
+            if(marc_output_file not in marc_output_files):
+                marc_output_files.append(marc_output_file)
+            
+        counter += 1
+        try:
+            bnb_yaz_client.sendline("find @attr 1=48 " + '"' + bnb_id + '"')
+            bnb_yaz_client.expect("Number of hits:.*", timeout=1000)
+            count_search = re.search(b"Number of hits: (\\d+), setno", bnb_yaz_client.after)
+            
+            if the_max_records_per_batch_for_delivery > 0:
+                if count_search.group(1).decode("utf-8") != "0":
+                    bnb_yaz_client.sendline("show all")
+                    bnb_yaz_client.expect("\r\n")
+
+                    number_of_downloaded_ids += 1
+                    new_bnb_info[source_file]["list_of_downloaded_id"].append(bnb_id)
+                    UpdateDateDeliveredInRecords(bnb_id)
+                    # if the maximum number of records per batch for delivery is reached, reset it and create a new marc output file
+                    the_max_records_per_batch_for_delivery -= 1
```

Comment:
The max should never change, please compare against the number_of_downloaded_ids